File: accounts-go/account.go
package account

import (
	"errors"
	"fmt"
)

type Account struct {
	id      *uuid
	ownerId *uuid
	balance int64
	open    bool
}

func (a *Account) Open(accountId, ownerId uuid) (Event, error) {
	if a.id != nil || a.ownerId != nil {
		return nil, errors.New("Account already open")
	}

	event := AccountOpenedEvent{accountId, ownerId}
	a.applyAccountOpened(event)
	return event, nil
}

func (a *Account) deposit(amount int64) (Event, error) {

	event := MoneyDepositedEvent{amount, a.balance + amount}
	a.applyMoneyDeposited(event)
	return event, nil
}

func (a *Account) applyAccountOpened(event AccountOpenedEvent) {
	a.id = &event.accountId
	a.ownerId = &event.ownerId
	a.balance = 0
	a.open = true
}

func (a *Account) applyMoneyDeposited(event MoneyDepositedEvent) {
	a.balance = event.balance
}

type Event interface {
	//Serialize() []byte
}

type AccountOpenedEvent struct {
	accountId uuid
	ownerId   uuid
}

type MoneyDepositedEvent struct {
	amountDeposited int64
	balance         int64
}

func (e AccountOpenedEvent) String() string {
	return fmt.Sprintf("AccountOpenedEvent{ownerId: %s}", e.ownerId)
}

/*func (e AccountOpenedEvent) Serialize() []byte {
	return e.ownerId
}*/

File: account-service/src/main/java/lt/rieske/accounts/domain/Account.java
package lt.rieske.accounts.domain;

import lt.rieske.accounts.eventsourcing.EventStream;
import lt.rieske.accounts.eventsourcing.EventVisitor;

import java.util.UUID;


public class Account implements EventVisitor<AccountEvent> {
    private final EventStream<Account, AccountEvent> eventStream;

    private final UUID accountId;

    private UUID ownerId;
    private long balance;
    private boolean open;

    public Account(EventStream<Account, AccountEvent> eventStream, UUID accountId) {
        this.eventStream = eventStream;
        this.accountId = accountId;
    }

    public AccountEvent.AccountSnapshot snapshot() {
        return new AccountEvent.AccountSnapshot(accountId, ownerId, balance, open);
    }

    public void open(UUID ownerId) {
        if (this.ownerId != null) {
            throw new IllegalStateException("Account already has an owner");
        }
        eventStream.append(new AccountEvent.AccountOpenedEvent(ownerId), this, accountId);
    }

    public void deposit(long amount) {
        if (amount == 0) {
            return;
        }
        requireOpenAccount();
        if (amount < 0) {
            throw new IllegalArgumentException("Can not deposit negative amount: " + amount);
        }
        eventStream.append(new AccountEvent.MoneyDepositedEvent(amount, balance + amount), this, accountId);
    }

    public void withdraw(long amount) {
        if (amount == 0) {
            return;
        }
        requireOpenAccount();
        if (amount < 0) {
            throw new IllegalArgumentException("Can not withdraw negative amount: " + amount);
        }
        if (balance < amount) {
            throw new IllegalArgumentException("Insufficient balance");
        }
        eventStream.append(new AccountEvent.MoneyWithdrawnEvent(amount, balance - amount), this, accountId);
    }

    public void close() {
        if (balance != 0) {
            throw new IllegalStateException("Balance outstanding");
        }
        eventStream.append(new AccountEvent.AccountClosedEvent(), this, accountId);
    }

    public UUID id() {
        return accountId;
    }

    public UUID ownerId() {
        return ownerId;
    }

    public long balance() {
        return balance;
    }

    boolean isOpen() {
        return open;
    }

    private void requireOpenAccount() {
        if (!open) {
            throw new IllegalStateException("Account not open");
        }
    }

    @Override
    public void visit(AccountEvent event) {
        switch (event) {
            case AccountEvent.AccountSnapshot snapshot -> {
                this.ownerId = snapshot.ownerId();
                this.balance = snapshot.balance();
                this.open = snapshot.open();
            }
            case AccountEvent.AccountOpenedEvent accountOpened -> {
                this.ownerId = accountOpened.ownerId();
                this.balance = 0;
                this.open = true;
            }
            case AccountEvent.AccountClosedEvent accountClosed -> this.open = false;
            case AccountEvent.MoneyDepositedEvent moneyDeposited -> this.balance = moneyDeposited.balance();
            case AccountEvent.MoneyWithdrawnEvent moneyWithdrawn -> this.balance = moneyWithdrawn.balance();
        }
    }
}
